src/socket/haven_socket.rs

```diff
@@ -1,52 +1,85 @@
-use std::time::Duration;
-
 use bytes::Bytes;
+use clone_macro::clone;
 use earendil_crypt::{Fingerprint, IdentitySecret};
 use earendil_packet::{crypt::OnionSecret, Dock};
-use smol::{Task, Timer};
+use moka::sync::Cache;
+use smol::{
+    channel::{Receiver, Sender},
+    Task, Timer,
+};
 use smol_timeout::TimeoutExt;
-use stdcode::StdcodeSerializeExt;
+use smolscale::immortal::{Immortal, RespawnStrategy};
+use std::time::Duration;
 
 use crate::{
     daemon::context::DaemonContext,
     global_rpc::{transport::GlobalRpcTransport, GlobalRpcClient},
-    haven::{HavenLocator, RegisterHavenReq, HAVEN_FORWARD_DOCK},
+    haven::{HavenLocator, RegisterHavenReq},
 };
 
-use super::{n2r_socket::N2rSocket, Endpoint, SocketRecvError, SocketSendError};
+use super::{
+    encrypter::{Encrypter, HavenMsg},
+    n2r_socket::N2rSocket,
+    Endpoint, SocketRecvError, SocketSendError,
+};
 
 pub struct HavenSocket {
     ctx: DaemonContext,
     n2r_socket: N2rSocket,
     identity_sk: IdentitySecret,
-    onion_sk: OnionSecret,
     rendezvous_point: Option<Fingerprint>,
-    _task: Option<Task<()>>,
+    _register_haven_task: Option<Task<()>>,
+    /// mapping between destination endpoints and encrypters
+    encrypters: Cache<Endpoint, Encrypter>,
+    /// buffer for decrypted incoming messages
+    recv_incoming_decrypted: Receiver<(Bytes, Endpoint)>,
+    send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
```

Comment:
We do not need the sending end of this channel here, right?

Reply:
This is so that `HavenSocket::send_to()` can use the sending end when it creates new encrypters